File: ➡️📂Core kernel Yaiwes/LangChain-Task-Steering/packages/typescript/src/skills.ts
/**
 * Skill loading utilities for task-scoped skills.
 *
 * Handles reading SKILL.md files from a backend and parsing their
 * YAML frontmatter into {@link SkillMetadata} objects.  Isolated from
 * `middleware.ts` to keep I/O and parsing separate from orchestration.
 */

import type { SkillMetadata } from './types.js'

const FRONTMATTER_RE = /^---\s*\n([\s\S]*?)\n---\s*\n/
const MAX_SKILL_FILE_SIZE = 10 * 1024 * 1024 // 10 MB

// ── Simple YAML frontmatter parser ─────────────────────────
// Handles the limited subset used by SKILL.md files:
// - Top-level scalar key-value pairs
// - Block sequences (- item)
// - One-level nested mappings (key:\n  k: v)

/** @internal */
export function parseSimpleYaml(text: string): Record<string, unknown> | null {
  const result: Record<string, unknown> = {}
  const lines = text.split('\n')
  let i = 0

  // Reject non-mapping top-level (e.g. starts with "- ")
  const firstNonEmpty = lines.find((l) => l.trim() !== '' && !l.trim().startsWith('#'))
  if (firstNonEmpty && /^\s*-\s/.test(firstNonEmpty)) {
    return null
  }

  while (i < lines.length) {
    const line = lines[i]

    // Skip empty lines and comments
    if (line.trim() === '' || line.trim().startsWith('#')) {
      i++
      continue
    }

    // Top-level key: value
    const kvMatch = line.match(/^([A-Za-z0-9_-]+)\s*:\s*(.*)/)
    if (!kvMatch) {
      // Unrecognised line at top level — likely malformed YAML
      i++
      continue
    }

    const key = kvMatch[1]
    const inlineValue = kvMatch[2].trim()

    if (inlineValue) {
      result[key] = inlineValue
      i++
    } else {
      // Look ahead for block content (list or nested mapping)
      i++
      const items: string[] = []
      const mapping: Record<string, string> = {}
      let isList = false
      let isMapping = false

      while (i < lines.length) {
        const nextLine = lines[i]
        if (nextLine.trim() === '' || nextLine.trim().startsWith('#')) {
          i++
          continue
        }

        // Not indented → back to top level
        if (!/^\s/.test(nextLine)) break

        const listItemMatch = nextLine.match(/^\s+- (.+)/)
        const nestedKvMatch = nextLine.match(/^\s+([A-Za-z0-9_-]+)\s*:\s*(.+)/)

        if (listItemMatch) {
          isList = true
          items.push(listItemMatch[1].trim())
          i++
        } else if (nestedKvMatch) {
          isMapping = true
          mapping[nestedKvMatch[1]] = nestedKvMatch[2].trim()
          i++
        } else {
          break
        }
      }

      if (isList) {
        result[key] = items
      } else if (isMapping) {
        result[key] = mapping
      }
      // If neither, key has no value — omit from result
    }
  }

  return Object.keys(result).length > 0 ? result : null
}

// ── Frontmatter parsing ────────────────────────────────────

/**
 * Parse YAML frontmatter from a SKILL.md file's content.
 *
 * Returns a `SkillMetadata` object on success, or `null` if the
 * content is invalid or missing required fields.
 */
export function parseSkillFrontmatter(content: string, path: string): SkillMetadata | null {
  if (content.length > MAX_SKILL_FILE_SIZE) {
    console.warn(`[langchain-task-steering] Skipping ${path}: file exceeds 10 MB size limit`)
    return null
  }

  const match = content.match(FRONTMATTER_RE)
  if (!match) {
    console.warn(`[langchain-task-steering] Skipping ${path}: no valid YAML frontmatter found`)
    return null
  }

  const data = parseSimpleYaml(match[1])
  if (!data) {
    console.warn(`[langchain-task-steering] Skipping ${path}: frontmatter is not a mapping`)
    return null
  }

  const name = String(data.name ?? '').trim()
  const description = String(data.description ?? '').trim()

  if (!name || !description) {
    console.warn(
      `[langchain-task-steering] Skipping ${path}: missing required 'name' or 'description'`
    )
    return null
  }

  const result: SkillMetadata = { name, description, path }

  const rawTools = data['allowed-tools']
  if (typeof rawTools === 'string') {
    result.allowedTools = rawTools.split(/\s+/)
  } else if (Array.isArray(rawTools)) {
    result.allowedTools = rawTools.map(String)
  }

  const license = String(data.license ?? '').trim() || null
  if (license) result.license = license

  const compatibility = String(data.compatibility ?? '').trim() || null
  if (compatibility) result.compatibility = compatibility

  const rawMetadata = data.metadata
  if (typeof rawMetadata === 'object' && rawMetadata !== null && !Array.isArray(rawMetadata)) {
    const metadataDict: Record<string, string> = {}
    for (const [k, v] of Object.entries(rawMetadata)) {
      metadataDict[String(k)] = String(v)
    }
    if (Object.keys(metadataDict).length > 0) {
      result.metadata = metadataDict
    }
  }

  return result
}

// ── Backend loading ────────────────────────────────────────

/**
 * Load skill metadata from backend source paths.
 *
 * For each source path, lists directories, constructs SKILL.md paths,
 * batch-downloads them, and parses frontmatter.  Mirrors the loading
 * flow in deepagents' `SkillsMiddleware`.
 *
 * The backend is duck-typed — it must provide:
 * - `ls(path)` → `{ entries: [...] }` or `[...]`
 * - `downloadFiles(paths)` or `download_files(paths)` → `[{ content, error }, ...]`
 */
export function loadSkillsFromBackend(backend: unknown, sourcePaths: string[]): SkillMetadata[] {
  const allSkills = new Map<string, SkillMetadata>()
  const b = backend as Record<string, unknown>

  for (const sourcePath of sourcePaths) {
    let lsResult: unknown
    try {
      lsResult = (b.ls as (p: string) => unknown)(sourcePath)
    } catch (exc) {
      console.warn(`[langchain-task-steering] Failed to list skill source '${sourcePath}': ${exc}`)
      continue
    }

    let entries: unknown[]
    const lsObj = lsResult as Record<string, unknown> | unknown[] | null
    if (Array.isArray(lsObj)) {
      entries = lsObj
    } else if (lsObj && typeof lsObj === 'object' && 'entries' in lsObj) {
      entries = (
        Array.isArray((lsObj as Record<string, unknown>).entries)
          ? (lsObj as Record<string, unknown>).entries
          : []
      ) as unknown[]
    } else {
      entries = []
    }

    const skillDirs: string[] = []
    for (const entry of entries) {
      if (typeof entry !== 'object' || entry === null) continue
      const e = entry as Record<string, unknown>
      const isDir = e.is_dir ?? e.isDir ?? false
      if (!isDir) continue
      const pathVal = e.path as string | undefined
      if (pathVal) skillDirs.push(pathVal)
    }

    if (skillDirs.length === 0) continue

    const mdPaths: Array<[string, string]> = skillDirs.map((dirPath) => {
      const mdPath = dirPath.endsWith('/') ? `${dirPath}SKILL.md` : `${dirPath}/SKILL.md`
      return [dirPath, mdPath]
    })

    const pathsToDownload = mdPaths.map(([, md]) => md)
    const downloadFn = (b.downloadFiles ?? b.download_files) as
      | ((paths: string[]) => unknown[])
      | undefined

    if (!downloadFn) {
      console.warn(
        `[langchain-task-steering] Backend does not have downloadFiles or download_files method`
      )
      continue
    }

    let responses: unknown[]
    try {
      responses = downloadFn.call(backend, pathsToDownload)
    } catch (exc) {
      console.warn(
        `[langchain-task-steering] Failed to download SKILL.md files from '${sourcePath}': ${exc}`
      )
      continue
    }

    for (let i = 0; i < mdPaths.length; i++) {
      const [, mdPath] = mdPaths[i]
      const response = responses[i] as Record<string, unknown> | null | undefined

      if (!response) continue
      if (response.error) continue

      const contentRaw = response.content
      if (contentRaw == null) continue

      let content: string
      if (typeof contentRaw === 'string') {
        content = contentRaw
      } else if (contentRaw instanceof Uint8Array) {
        content = new TextDecoder().decode(contentRaw)
      } else {
        content = String(contentRaw)
      }

      const skill = parseSkillFrontmatter(content, mdPath)
      if (skill) {
        // Last source wins on name conflict
        allSkills.set(skill.name, skill)
      }
    }
  }

  return Array.from(allSkills.values())
}

File: ➡️📂Core kernel Yaiwes/LangChain-Task-Steering/packages/python/src/langchain_task_steering/_skills.py
"""Skill loading utilities for task-scoped skills.

Handles reading SKILL.md files from a backend and parsing their
YAML frontmatter into ``SkillMetadata`` dicts.  Isolated from
``middleware.py`` to keep I/O and parsing separate from orchestration.
"""

import logging
import re
from pathlib import PurePosixPath
from typing import Any

from .types import SkillMetadata

logger = logging.getLogger(__name__)

_FRONTMATTER_RE = re.compile(r"^---\s*\n(.*?)\n---\s*\n", re.DOTALL)
_MAX_SKILL_FILE_SIZE = 10 * 1024 * 1024  # 10 MB


_yaml_module = None


def _lazy_yaml():
    """Lazy-import pyyaml with a helpful error if not installed."""
    global _yaml_module
    if _yaml_module is not None:
        return _yaml_module
    try:
        import yaml

        _yaml_module = yaml
        return yaml
    except ImportError:
        raise ImportError(
            "pyyaml is required for skill loading. "
            "Install it with: pip install pyyaml  "
            "(or: pip install langchain-task-steering[skills])"
        ) from None


def parse_skill_frontmatter(content: str, path: str) -> SkillMetadata | None:
    """Parse YAML frontmatter from a SKILL.md file's content.

    Returns a ``SkillMetadata`` dict on success, or ``None`` if the
    content is invalid or missing required fields.
    """
    if len(content) > _MAX_SKILL_FILE_SIZE:
        logger.warning("Skipping %s: file exceeds 10 MB size limit", path)
        return None

    match = _FRONTMATTER_RE.match(content)
    if not match:
        logger.warning("Skipping %s: no valid YAML frontmatter found", path)
        return None

    yaml = _lazy_yaml()
    try:
        data = yaml.safe_load(match.group(1))
    except yaml.YAMLError as exc:
        logger.warning("Invalid YAML in %s: %s", path, exc)
        return None

    if not isinstance(data, dict):
        logger.warning("Skipping %s: frontmatter is not a mapping", path)
        return None

    name = str(data.get("name", "")).strip()
    description = str(data.get("description", "")).strip()

    if not name or not description:
        logger.warning("Skipping %s: missing required 'name' or 'description'", path)
        return None

    # Optional fields
    raw_tools = data.get("allowed-tools")
    if isinstance(raw_tools, str):
        allowed_tools = raw_tools.split()
    elif isinstance(raw_tools, list):
        allowed_tools = [str(t) for t in raw_tools]
    else:
        allowed_tools = []

    license_str = str(data.get("license") or "").strip() or None
    compat_str = str(data.get("compatibility") or "").strip() or None

    raw_metadata = data.get("metadata", {})
    metadata_dict: dict[str, str] = {}
    if isinstance(raw_metadata, dict):
        metadata_dict = {str(k): str(v) for k, v in raw_metadata.items()}

    result: SkillMetadata = {
        "name": name,
        "description": description,
        "path": path,
    }
    if license_str is not None:
        result["license"] = license_str
    if compat_str is not None:
        result["compatibility"] = compat_str
    if metadata_dict:
        result["metadata"] = metadata_dict
    if allowed_tools:
        result["allowed_tools"] = allowed_tools

    return result


def load_skills_from_backend(
    backend: Any, source_paths: list[str]
) -> list[SkillMetadata]:
    """Load skill metadata from backend source paths.

    For each source path, lists directories, constructs SKILL.md paths,
    batch-downloads them, and parses frontmatter.  Mirrors the loading
    flow in deepagents' ``SkillsMiddleware`` but uses our own types.
    """
    all_skills: dict[str, SkillMetadata] = {}

    for source_path in source_paths:
        try:
            ls_result = backend.ls(source_path)
        except Exception as exc:
            logger.warning("Failed to list skill source '%s': %s", source_path, exc)
            continue

        # LsResult has .entries, but duck-type for flexibility
        entries = getattr(ls_result, "entries", None)
        if entries is None:
            entries = ls_result if isinstance(ls_result, list) else []

        # Collect subdirectories (duck-type: dict or object with .is_dir/.path)
        skill_dirs: list[str] = []
        for entry in entries or []:
            is_dir = (
                entry.get("is_dir")
                if isinstance(entry, dict)
                else getattr(entry, "is_dir", False)
            )
            if not is_dir:
                continue
            path_val = (
                entry["path"]
                if isinstance(entry, dict)
                else getattr(entry, "path", None)
            )
            if path_val:
                skill_dirs.append(path_val)

        if not skill_dirs:
            continue

        # Build SKILL.md paths
        md_paths: list[tuple[str, str]] = []
        for dir_path in skill_dirs:
            md_path = str(PurePosixPath(dir_path) / "SKILL.md")
            md_paths.append((dir_path, md_path))

        # Batch download
        paths_to_download = [md for _, md in md_paths]
        try:
            responses = backend.download_files(paths_to_download)
        except Exception as exc:
            logger.warning(
                "Failed to download SKILL.md files from '%s': %s",
                source_path,
                exc,
            )
            continue

        # Parse each response
        for (dir_path, md_path), response in zip(md_paths, responses, strict=True):
            error = getattr(response, "error", None)
            if error:
                continue

            content_bytes = getattr(response, "content", None)
            if content_bytes is None:
                continue

            try:
                content = (
                    content_bytes.decode("utf-8")
                    if isinstance(content_bytes, bytes)
                    else str(content_bytes)
                )
            except UnicodeDecodeError as exc:
                logger.warning("Error decoding %s: %s", md_path, exc)
                continue

            skill = parse_skill_frontmatter(content, md_path)
            if skill:
                # Last source wins on name conflict
                all_skills[skill["name"]] = skill

    return list(all_skills.values())
